tests/Operations/Merge.hs

```diff
@@ -0,0 +1,157 @@
+        [("L", DI.fromList [10 :: Int, 20])]
+
+dfSide2 :: D.DataFrame
+dfSide2 =
+    D.fromNamedColumns
+        [("R", DI.fromList ["x" :: String, "y"])]
+
+mergeVerticalWAI :: Test
+mergeVerticalWAI =
+    TestCase
+        ( assertEqual
+            "Vertical merge concatenates matching columns"
+            ( D.fromNamedColumns
+                [ ("A", DI.fromList [1 :: Int, 2, 3])
+                , ("B", DI.fromList ['a', 'b', 'c'])
+                ]
+            )
+            (df1 <> df2)
+        )
+
+mergeVerticalDisjointSchemas :: Test
+mergeVerticalDisjointSchemas =
+    TestCase
+        ( assertEqual
+            "Disjoint schemas produce optional padding"
+            ( D.fromNamedColumns
+                [
+                    ( "A"
+                    , DI.fromList
+                        [ Just (1 :: Int)
+                        , Just 2
+                        , Nothing
+                        , Nothing
+                        ]
+                    )
+                ,
+                    ( "B"
+                    , DI.fromList
+                        [ Nothing
+                        , Nothing
+                        , Just 'c'
+                        , Just 'd'
+                        ]
+                    )
+                ]
+            )
+            (dfOnlyA <> dfOnlyB)
+        )
+
+mergeVerticalWithEmptyLeft :: Test
+mergeVerticalWithEmptyLeft =
+    TestCase
+        ( assertEqual
+            "Empty <> df = df"
+            df1
+            (D.empty <> df1)
+        )
+
+mergeVerticalWithEmptyRight :: Test
+mergeVerticalWithEmptyRight =
+    TestCase
+        ( assertEqual
+            "df <> Empty = df"
+            df1
+            (df1 <> D.empty)
+        )
+
+mergeHorizontalWAI :: Test
+mergeHorizontalWAI =
+    TestCase
+        ( assertEqual
+            "Horizontal merge combines columns side by side"
+            ( D.fromNamedColumns
+                [ ("L", DI.fromList [10 :: Int, 20])
+                , ("R", DI.fromList ["x" :: String, "y"])
+                ]
+            )
+            (dfSide1 ||| dfSide2)
+        )
+
+mergeHorizontalWithEmptyRight :: Test
+mergeHorizontalWithEmptyRight =
+    TestCase
+        ( assertEqual
+            "df ||| empty = df"
+            dfSide1
+            (dfSide1 ||| D.empty)
+        )
+
+mergeHorizontalWithEmptyLeft :: Test
+mergeHorizontalWithEmptyLeft =
+    TestCase
+        ( assertEqual
+            "empty ||| df = df"
+            dfSide2
+            (D.empty ||| dfSide2)
+        )
+
+mergeVerticalDifferentTypesSameColumnName :: Test
+mergeVerticalDifferentTypesSameColumnName =
+    let
+        dfIntA = D.fromNamedColumns [("A", DI.fromList [1 :: Int, 2])]
+
+        dfTextA = D.fromNamedColumns [("A", DI.fromList ["x" :: String, "y"])]
+     in
+        TestCase
+            ( assertEqual
+                "Merging columns with same name but different types drops the column (current behavior)"
+                D.empty
```

Comment:
It's not clear to me what should happen in this case. It seems that such columns are just not included  in the merged dataframe.

Reply:
I really like the `Either` approach.